flink-runtime/src/main/java/org/apache/flink/runtime/sink/coordinator/SinkCoordinator.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.sink.coordinator;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.runtime.jobgraph.OperatorID;
+import org.apache.flink.runtime.operators.coordination.CoordinatorStore;
+import org.apache.flink.runtime.operators.coordination.OperatorCoordinator;
+import org.apache.flink.runtime.operators.coordination.OperatorEvent;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * The default implementation of the {@link OperatorCoordinator} for the {@link
+ * org.apache.flink.api.connector.sink2.Sink}.
+ *
+ * <p>The <code>SinkCoordinator</code> helps trigger an immediate global checkpoint when all sink
+ * operators are reaching end of data.
+ */
+@Internal
+public class SinkCoordinator implements OperatorCoordinator {
```

Comment:
I have a couple of remarks about this construct:
- What if some parts of a job don't have any sink? 
- What will happen if Sink has it's own `Coordinator`? Can we have two coordinators at the same time?
- I think this is pretty invasive design. We need to send extra events from the operators to JM, we need to expose triggering checkpoints to the coordinators, we have to wire everything. This touches, modifies and connects a lot of components. Why can not all of this happen purely in the JM, maybe just inside `CheckpointCoordinator`? We could add another `ExecutionState.WAITING_FOR_FINAL_CHECKPOINT` or maybe simply `FINISHING`, similarly how we added `ExecutionState.INITIALIZING` not that long time ago and just listen in the JM (or `CheckpointCoordinator`) until all subtasks have reached this new state?

Reply:
The implementation has been revised and the above concerns are all resolved.

Reply:
I'm not sure if this was really resolved; we should have a discussion on whether a dedicated execution state wouldn't be the better option, or rather what the pros/cons are of each approach.

My impression is that we're de-facto introducing a new execution state, and wire a separate path through the entire stack.
If there were a `WAITING_FOR_FINAL_CHECKPOINT` we could just re-use the existing path for updates. This would clarify to the runtime that this state actually exists (and has to be accounted for!), in what order things happen, how often this event occurs.

Reply:
@zentol Jiangxin and I discussed the approach of adding a dedicated execution state (e.g. FINISHING). We chose the current approach because it is less invasive and simpler to implement. 

For example, given that FINISHING effectively belongs to what is currently RUNNING, we would need to go through the code (47+ places) that currently reference ExecutionState#RUNNING and check if any change is needed. This would introduce more complexity and make the PR error prone.

I am not sure if there is extra benefit of "clarify to the runtime that this state actually exists" in addition to what this PR does. Can you help explain what is the benefit of introducing an extra execution state?

Reply:
> How do you know that no part of the system is doing something that shouldn't happen after the task started to wait for the checkpoint? If we just closed our eyes and assumed there to not be a problem, well then that is quite a red flag.

I am sure we can always find more benefits of doing some extra work. The question is whether the benefit outweighs the cost (e.g. extra code complexity) and whether its priority is higher (for both the community and the volunteer who offered to improve Flink open source code).

The scope of FLINK-28386 is to prevent tasks having to wait for a checkpoint. While I agree it might beneficial for the adaptive scheduler to not scale up the job, and to provide metrics capturing how much time operators spend idling waiting for that next checkpoint, this is strictly speaking out of the scope of this PR.

As far as FLINK-28386's goal is concerned, do you see a better approach to achieve this goal, or do you see any performance or correctness issue with the current PR?

Reply:
I hope we can collaborate together in a friendly/reasonable way to improve Flink. And hopefully we can fix this issue in Flink 1.18 (which will code freeze on July 11) so that users can avoid waiting for checkpoint sooner than later.

If there is any performance or correctness issue, or if there is a simpler and more maintainable way to address FLINK-28386, I totally agree we should use that approach for long term benefits. In this particular case, I think the extra benefits around adding an extra ExecutionState can be worked on independently after this PR.

Reply:
@zentol Do you want to take another look?

For those features (e.g. metrics) that are not necessarily tied to the FLINK-28386 (i.e. trigger a checkpoint immediately after all tasks finish), can we agree to create a JIRA and work on that separately?